TTS/tts/models/naturalspeech2.py

```diff
@@ -0,0 +1,1223 @@
+
+        alignment_mas = maximum_path(
+            alignment_soft.squeeze(1).transpose(1, 2).contiguous(), attn_mask.squeeze(1).contiguous()
+        )
+        alignment_hard = torch.sum(alignment_mas, -1).int()
+        alignment_soft = alignment_soft.squeeze(1).transpose(1, 2)
+        return alignment_hard, alignment_soft, alignment_logprob, alignment_mas
+
+    @staticmethod
+    def _set_cond_input(aux_input: Dict):
+        # [TODO] use this
+        return None
+
+    def forward(  # pylint: disable=dangerous-default-value
+        self,
+        tokens: torch.tensor,
+        tokens_lens: torch.tensor,
+        latents: torch.tensor,
+        latents_lengths: torch.tensor,
+        mel: torch.tensor,
+        mel_lens: torch.tensor,
+        pitch: torch.tensor,
+        aux_input={"prompt": None, "durations": None, "language_ids": None},
+    ) -> Dict:
+        outputs = {}
+        tokens_mask = torch.unsqueeze(sequence_mask(tokens_lens, tokens.shape[1]), 1).float()
+        phoneme_enc = self.phoneme_encoder(tokens.unsqueeze(2))
+
+        latent_lens = torch.tensor(latents.shape[1:2]).to(latents.device)
+        mel_mask = torch.unsqueeze(sequence_mask(mel_lens, None), 1).float()
+        latents_mask = torch.unsqueeze(sequence_mask(latent_lens, None), 1).float()
+        # create masks for the segments and remaining parts
+        remaining_mask = torch.ones_like(latents, dtype=torch.bool)
+
+        # Get random segment for the speech prompt
+        speech_prompts, segment_indices = rand_segments(
+            latents, latents_lengths, self.diff_segment_size, let_short_samples=True, pad_short=True
+        )
+
+        # iterate over the batch dimension
+        for i in range(latents.size(0)):
+            remaining_mask[i, :, segment_indices[i] : segment_indices[i] + self.diff_segment_size] = 0
+
+        # apply masks to get remaining parts
+        remaining_latents = latents.masked_select(remaining_mask).view(latents.shape[0], latents.shape[1], -1)
+        remaining_latents_lengths = torch.tensor(remaining_latents.shape[1:2]).to(remaining_latents.device)
+
+        # Encode speech prompt
+        speech_prompts_enc = self.prompt_encoder(speech_prompts)
+
+        alignment_hard, alignment_soft, alignment_logprob, alignment_mas = self._forward_aligner(
+            phoneme_enc.transpose(1, 2), mel.transpose(1, 2), tokens_mask, mel_mask
+        )
+
+        alignment_soft = alignment_soft.transpose(1, 2)
+        alignment_mas = alignment_mas.transpose(1, 2)
+
+        durations_pred = self.duration_predictor(phoneme_enc.transpose(1, 2), speech_prompts_enc.transpose(1, 2))
+        durations_pred = durations_pred.squeeze(1)
+
+        pitch = average_over_durations(pitch, alignment_hard)
+        pitch_pred = self.pitch_predictor(phoneme_enc.transpose(1, 2), speech_prompts_enc.transpose(1, 2))
+
+        durations_pred_exp = durations_pred.unsqueeze(2)
+        expanded_dur = durations_pred_exp.expand_as(phoneme_enc)
```

Comment:
shouldn't we expanding the `phoneme_enc` not `duraitons_pred`

Reply:
duration pred shape is b , t , 1 and phoneme_enc size is b,t , 512 so here matching the size to match the dumentions, how can i expand phoneme_enc based on duration is not clear to me

Reply:
this is what we do in VITS 

```python
from TTS.tts.utils.helpers import generate_path

....

# convert duraitons to attn mask 
attn_mask = x_mask * y_mask.transpose(1, 2)  # [B, 1, T_enc] * [B, T_dec, 1]
attn = generate_path(durations.squeeze(1), attn_mask.squeeze(1).transpose(1, 2))

...

 # expand prior
m_p = torch.einsum("klmn, kjm -> kjn", [attn, m_p])
logs_p = torch.einsum("klmn, kjm -> kjn", [attn, logs_p])
```